botocore/stub.py

```diff
@@ -0,0 +1,201 @@
+
+        :param method: The name of the service method to return the error on.
+        :type method: str
+
+        :param service_error_code: The service error code to return,
+                                   e.g. NoSuchBucket
+        :type service_error_code: str
+
+        :param service_message: The service message to return, e.g.
+                        'The specified bucket does not exist.'
+        :type service_message: str
+
+        :param http_status_code: The HTTP status code to return, e.g. 404, etc
+        :type http_status_code: int
+        """
+        http_response = Response()
+        http_response.status_code = http_status_code
+
+        # We don't look to the model to build this because the caller would need
+        # to know the details of what the HTTP body would need to look like.
+        parsed_response = {
+            'ResponseMetadata': {'HTTPStatusCode': http_status_code},
+            'Error': {
+                'Message': service_message,
+                'Code': service_error_code
+            }
+        }
+
+        operation_name = self.client.meta.method_to_api_mapping.get(method)
+        response = (operation_name, (http_response, parsed_response))
+        self._queue.append(response)
+
+    def assert_no_pending_responses(self):
+        """
+        Asserts that all expected calls were made.
+        """
+        remaining = len(self._queue)
+        if remaining != 0:
+            raise AssertionError("%d responses remaining in queue." % remaining)
+
+    def _get_response_handler(self, model, params, **kwargs):
+        if not self._queue:
+            raise StubResponseError(
+                operation_name=model.name,
+                reason='Unexpected API Call: called with parameters %s' %
+                       params)
+
+        name, response = self._queue.popleft()
+        if name != model.name:
+            raise StubResponseError(
+                operation_name=model.name,
+                reason='Operation mismatch: found response for %s.' % name)
+
+        return response
+
+    def _validate_response(self, operation_name, service_response):
+        service_model = self.client.meta.service_model
+        operation_model = service_model.operation_model(operation_name)
+        output_shape = operation_model.output_shape
+
+        # Remove ResponseMetadata so that the validator doesn't attempt to
+        # perform validation on it.
+        response = service_response
+        if 'ResponseMetadata' in response:
+            response = copy.copy(service_response)
+            del response['ResponseMetadata']
+
+        if output_shape is not None:
+            validate_parameters(response, output_shape)
+        elif response:
+            # If the output shape is None, that means the response should be
+            # empty apart from ResponseMetadata
+            raise ParamValidationError(
+                report="Service response contains unknown members.")
```

Comment:
This message is a bit ambiguous. Maybe say something that the response should have at most the ResponseMetadata member.